### PR TITLE
Deprecate Matrix constructors with `::Type{MatElem}` param

### PR DESCRIPTION
**Pull request by lgoettgens**
These seem old and shouldn't be needed downstream

### CODE REVIEW

**Review by fingolfin — APPROVED** [review]

File: src/Deprecations.jl [hunk 1]
```diff
@@ -21,3 +21,11 @@ end
 @deprecate factor(f::PolyRingElem, R::Field) factor(R, f)
 
 @deprecate roots(f::PolyRingElem, R::Field) roots(R, f)
+
+# Deprecated in 0.32.*
+
+@deprecate zero_matrix(::Type{MatElem}, R::Ring, n::Int) zero_matrix(R, n)
```
[the remark this answers is not in the document]
Comment (thofma) [reply src/Deprecations.jl:27]: Yes, exactly. What is the problem with these methods?

[the remark this answers is not in the document]
Comment (lgoettgens) [reply src/Deprecations.jl:27]: It induces weird stuff downstream. E.g. in Nemo `zero_matrix(MatElem, QQ, 3)` doesn't return a `MatElem` (which users think is a `Generic.MatElem` but a `QQMatrix`.

**Comment by thofma** [comment]
I need to write functions of the form
```
function f(MatrixType::Type{<:Any}, x, y)
  zero_matrix(MatrixType, x, y)
end
```
How do I do this if we remove these methods here?

**Comment by lgoettgens** [comment]
But if you really have this use-case, then the functions touched in this Pr should delegate to Generic.Mat, right?

After thinking again about it, your use-case does only make sense to me for concrete types. The functions here however, have an abstract type as the first argument

**Comment by thofma** [comment]
No, they should not delegate to `Generic.Mat`. And no, `x` and `y` do not determine the output type. (I have clarified the 

The idea is that there is a parameter, which decides whether sparse or dense matrices are used for the computation. E.g., if `R == QQ`, it should use `QQMatrix` if `T == MatElem` and `SMat{QQFieldElem}` if `T == SMat` and Hecke is loaded.

**Comment by thofma** [comment]
I am happy to hear suggestions how to do this differently. I agree that it is not optimal, that `MatElem` is not really the type of the output, but I am not sure what a good name is.

**Pull request closed without merging**